site-src/geps/gep-726.md

```diff
@@ -0,0 +1,270 @@
+// these filters may be used on a Route rule. This may not be used on the same
+// Route rule as a HTTPRequestRewrite filter.
+type HTTPRequestRedirect struct {
+  // Path defines a path redirect.
+  //
+  // Support: Extended
+  //
+  // +optional
+  Path *HTTPPathRedirect `json:"path,omitempty"`
+  // ...
+}
+
+// HTTPPathRedirectType defines the type of path redirect.
+type HTTPPathRedirectType string
+
+const (
+  // This type of redirect indicates that the complete path will be replaced by
+  // the path redirect value.
+  AbsoluteHTTPPathRedirect HTTPPathRedirectType = "Absolute"
+
+  // This type of redirect indicates that the prefix match will be replaced by
+  // the path redirect value. This type only applies to path matches of type
+  // Prefix. If a Route has other match types, this redirect MUST NOT be applied
+  // to them.
+  ReplacePathPrefixMatchHTTPPathRedirect HTTPPathRedirectType = "ReplacePathPrefixMatch"
+)
+
+// HTTPPathRedirect defines configuration for path redirects.
+type HTTPPathRedirect struct {
+  // Type defines the type of path redirect.
+  //
+  // +kubebuilder:validation:Enum=Absolute;Prefix
+  Type HTTPPathRedirectType `json:"type"`
+
+  // Value defines the HTTP path to redirect to.
+  //
+  // +kubebuilder:validation:MaxLength=1024
+  Value HTTPPathRedirectType `json:"value"`
+}
+```
+
+### Rewrites
+
+A new `RequestRewrite` filter can be added to support rewrites. In the following
+example, a a request to `example.com/foo/abc` would be rewritten to
+`example.net/bar/abc`.
+
+```yaml
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: http-filter-1
+spec:
+  rules:
+    - matches:
+      - path:
+          type: Prefix
+          value: /foo
+      filters:
+      - type: RequestRewrite
+        requestRewrite:
+          hostname: example.net
+          path:
+            type: ReplacePathPrefixMatch
+            value: /bar
+```
+
+This would be represent with the following API additions:
+```go
+// HTTPRequestRewrite defines a filter that modifies a request during
+// forwarding. At most one of these filters may be used on a Route rule. This
+// may not be used on the same Route rule as a HTTPRequestRedirect filter.
+type HTTPRequestRewrite struct {
+  // Hostname is the value to be used to replace the Host header value during
+  // forwarding.
+  //
+  // Support: Extended
+  //
+  // +optional
+  // +kubebuilder:validation:MaxLength=255
+  Hostname *string `json:"hostname,omitempty"`
+
+  // Path defines a path rewrite.
+  //
+  // Support: Extended
+  //
+  // +optional
+  Path *HTTPPathRewrite `json:"path,omitempty"`
+}
+
+// HTTPPathRewriteType defines the type of path rewrite.
+type HTTPPathRewriteType string
+
+const (
+  // This type of rewrite indicates that the prefix match will be replaced by
+  // the path rewrite value. This type only applies to path matches of type
+  // Prefix. If a Route has other match types, this rewrite MUST NOT be applied
```

Comment:
What if I want to route based on hostname and not path, but I still want to strip a part of the path?
So example.com/foo/bar/baz goes to internal.example.com/bar/baz? This is a very common use-case where users use the routing infrastructure to "fix" such things.

Reply:
I think this would still be possible though? The prefix match would be "foo" and the redirect/rewrite value would be "". I need to document that that's possible, but I don't think there's currently anything that would prevent that..

Reply:
Added some clarification here, may have overcomplicated the wording though.